File: Sem1-2/CPP_Java2024/List3/CPP/src/Rhombus.cpp
#include <Rhombus.hpp>
#include <utils.hpp>

#include <stdexcept>
#include <cmath>

Rhombus::Rhombus(double side, double angle) : Quad("Rhombus"), side(side), angle(angle)
{
    if(side < 0)
    {
        throw std::invalid_argument(LOG_LOC() + "Rhombus side cannot be negative, got: " + std::to_string(side));
    }
    if(angle < 0 || angle > 180)
    {
        throw std::invalid_argument(LOG_LOC() + "Rhombus angle can be neither negative nor reflex, got: " + std::to_string(angle));
    }
}

double Rhombus::Circumference() const
{
    return 4.0 * side;
}

double Rhombus::Area() const
{
    return side * side * std::sin(angle * (M_PI / 180.0));
}

File: Sem1-2/CPP_Java2024/List3/JAVA/Rhombus.java
public class Rhombus extends Quad
{
    protected final double side;
    protected final double angle;

    public Rhombus(double side, double angle)
    {
        super("Rhombus");
        if(side < 0)
        {
            throw new IllegalArgumentException("Rhombus side cannot be negative, got: " + side);
        }
        if(angle < 0 || angle > 180)
        {
            throw new IllegalArgumentException("Rhombus angle can be neither negative nor reflex, got: " + angle);
        }
        this.side = side;
        this.angle = angle;
    }

    @Override
    public double Circumference()
    {
        return 4.0 * side;
    }

    @Override
    public double Area()
    {
        return side * side * Math.sin(Math.toRadians(angle));
    }
}